### ISSUE TITLE
[JENKINS-9913] Not obvious why some post-build tasks enforce serial behavior even when builds are concurrent

### ISSUE THREAD

**Issue by timja-bot** [timja-org/jenkins-gh-issues-poc-11-04#7859]
<p>We're experiencing an issue with concurrent builds where Jenkins appears to be associating separate builds (run on different machines) such that they won't be marked as completed until all jobs are completed.  For example, if we kick off 5 concurrent builds on 5 different nodes, builds 1-4 won't be marked as completed if build #5 is still running, even though builds 1-4 are finished.  I've seen a report of someone experiencing this issue elsewhere:</p>

<p><a href="http://groups.google.com/group/jenkinsci-users/browse_thread/thread/e477e25910266d2a?fwc=1" class="external-link" target="_blank" rel="nofollow noopener">http://groups.google.com/group/jenkinsci-users/browse_thread/thread/e477e25910266d2a?fwc=1</a></p>

<p>but a solution wasn't posted.  We do not have the batch plugin or the locks and latches plugin installed.  We've disabled all post-build processing and switched between different containers (Glassfish/Tomcat), but the problem persists.  I couldn't find an issue logged for this other than the aforementioned posting.</p>

---
<details><summary><i>Originally reported by <a title="Philip Metting van Rijn" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pomvr">pomvr</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-9913" target="_blank">Not obvious why some post-build tasks enforce serial behavior even when builds are concurrent</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jesse Glick" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2013-08-06T22:09:21+00:00
<li><b>imported</b>: 2025-11-05
</ul></i>
</details>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=tiainpa">tiainpa</a>:</i>
<p>We have a similar issue: The build which should finish first gets stuck due to a bug in our test framework (last line in build console is 'Recording test results'), and all the concurrent builds running at the same time get stuck on the same phase, finishing only when the first build is forcefully killed.</p>

<p>This might have something to do with the JUnit test result report publishing?</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ahawtho">ahawtho</a>:</i>
<p>We see the same thing - we have a test job that can be triggered with a few different upstream jobs to specify different sets of tests to run with different test parameters.  We use the parameterized trigger plugin to kick off multiple concurrent instances of the downstream job.  One of our "trigger" jobs specifies a set of long running performance tests (about 10 hours).  Another is a regression test suite, which only takes a few minutes.  We've seen instances at least twice where one of the regression tests will start after the long-running performance tests.  The short-lived regression test job will remain running until the performance test completes, which can be ten hours later.  The console log shows the regression test job as completed, having archived and recorded fingerprints, but it's still marked as running.  Furthermore, an attempt to stop the regression test job caused the long-running performance test job to stop.  Several regression tests would normally be queued during this time.  This seems to have been recent in an upgrade to 1.448.  I'm not sure what we were running immediately before that.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dannystaple">dannystaple</a>:</i>
<p>We were hoping that disabling the Junit post build task would stop this - it didn't do so. This has been killing us for some test jobs where we use a concurrent job to massively reduce duplication of configuration (and all the headaches that come with that).</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mlewicka">mlewicka</a>:</i>
<p>A very unfortunate bug indeed, especially if one of the batch of jobs getting launched gets stuck for whatever reasons (remote execution over ssh, etc.). The stuck job will cause the whole batch to get stuck, and the possibility of this happening is exactly why we're using concurrent executions in first place...</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svs57">svs57</a>:</i>
<p>I am afraid that no one is working on a the investigation of this bug.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=xavier_leprevost">xavier_leprevost</a>:</i>
<p>It would be nice to know if someone will investigate this problem.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=inbar_rose">inbar_rose</a>:</i>
<p>same problem here. total blocker. task A starts, then task B starts. task B reaches the 'Recording test results' stage and hangs until task A finishes. after testing with simple timed builds with many plugins/options enabled/disabled concluded that junit is the problem. - found another issue like this here: <a href="https://issues.jenkins-ci.org/browse/JENKINS-10234" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-10234</a> </p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>Similar to <a href="https://issues.jenkins.io/browse/JENKINS-10234" title="Junit result archiver getting stuck for a long time in concurrent builds" class="issue-link" data-issue-key="JENKINS-10234"><del>JENKINS-10234</del></a> - for some causes even the same.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svs57">svs57</a>:</i>
<p>It is not a feature. All concurent builds execute in separated workspace and must  archive its artifacts independently. No need to wait others builds if current have finished.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>Sergey, do you mean 'artifact archiving' or 'JUnit result archiving'. In the former case, you would be probably right, but I've seen no comment here that artifact archiving is blocking, too.<br/>
In the case of JUnit: yes it MUST block to calculate the diff (regressions et.al.) to the previous test results. So it is a feature.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svs57">svs57</a>:</i>
<p>I mean all result of concurent job.<br/>
For example. I have Jenkins job that execute E2E tests on different QA servers. I started 10 jobs. One job running 5 hours. But others takes 1 hour. Then I can't see results of finished job and have to wait 4 hours to see it. We must have instrument to prevent this.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>For that you can e.g. use the xunit plugin as mentioned in <a href="https://issues.jenkins.io/browse/JENKINS-10234" title="Junit result archiver getting stuck for a long time in concurrent builds" class="issue-link" data-issue-key="JENKINS-10234"><del>JENKINS-10234</del></a>.</p>

<p>Generally, don't use any build steps which require blocking behaviour. Yes, very unfortunately it's not visible for end users which build steps do and which don't.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mantengamoslacalma">mantengamoslacalma</a>:</i>
<p>How is this a feature?</p>

<p>I do not use JUnit. I have a gerrit plugin that triggers a (parameterized) build+test job (a build script, and another script that runs some tests, all in a small shell snippet) whenever code is pushed to the repository. When the job is completed (successfully or not), an email notification needs to go out to the authors. It doesn't get much simpler than this.</p>

<p>Commits are independent from each other, and as a consequence, so are the build jobs. Why would you want the email plugin to sit in some checkpoint if the job is done?</p>

<p>Is there a workaround for this? This causes horrendous problems in our setup: aside from the resource waste (many nodes spend long periods of time waiting for slower builds), bugs in the test code will cause the whole cluster to deadlock.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=teh444">teh444</a>:</i>
<p>Happening in 1.515</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=surya548">surya548</a>:</i>
<p>>Exactly why separate branch projects are needed.</p>

<p>I dont understand what you suggest that I do now. Our Jenkins Instance is basically unusable because of these useless checkpoints.<br/>
I am not sure if it is practical  for me to open bugs against 9 plugins( that i know of, after looking at source code for all the plugins we use) and ask them to remove checkpoints. </p>

<p>Why can't we get rid of (or atleast make it optional) a obsolete feature designed for SVN/CVS . <br/>
As others have said in the comments above we would rather have stable Jenkins than some  optional feature than no-one cares about.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=surya548">surya548</a>:</i>
<p>>Applies equally to Git as to SVN; both support branches.</p>

<p>There is a big big difference, svn branches are long lived and most git branches typically are short lived( pull requests for example) . And having linear history doesn't make any sense for a branch with only a few commits.<br/>
The same goes for calculating changesets, culprits etc. None of those features are relevant for short lived branches. </p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Changelogs and the like are relevant if there is more than one commit in the branch, which is common in Git pull requests. And SVN branches need not be long lived since it is just as cheap to create and dispose of them as it is in Git. (CVS is a different matter of course.) Whether you care about linear build history depends on your workflow, not the SCM per se.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=surya548">surya548</a>:</i>
<p>> Whether you care about linear build history depends on your workflow, not the SCM per se.</p>

<p>As I've been stating over and over, if it depends on our workflow there should be an option to configure if we need linear history or not based on our workflow. Right now jenkins is assuming that everyone needs this linear history thing.</p>

<p>Your solution is to go though source code of each plugin and open bugs against them which is extremely impractical and it feels extremely silly to expect users to  go though plugins source code before installing it.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=surya548">surya548</a>:</i>
<p>>  more than one commit in the branch, which is common in Git pull requests</p>

<p>I am not sure how you came up with that (misguided) conclusion. <br/>
Here are the number of commits on last few  open pull requests on jenkinci/jenkins . As you can see most of them are short lived with 1 or 2 commits.</p>


<p>917 - 1<br/>
916 - 2<br/>
914 - 1<br/>
913 - 1<br/>
912 - 3<br/>
911 - 1<br/>
910 - 1<br/>
907 - 1<br/>
904 - 2<br/>
902 - 1<br/>
894 - 1<br/>
891 - 1<br/>
889 - 2<br/>
884 - 1<br/>
882 - 1<br/>
877 - 2<br/>
858 - 3<br/>
855 - 2<br/>
847 - 2<br/>
840 - 4<br/>
839 - 1<br/>
823 - 4<br/>
816 - 7<br/>
808 - 1<br/>
806 - 1<br/>
799 - 1<br/>
782 - 1<br/>
776 - 1</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Around half of those have more than one commit, as I said.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=surya548">surya548</a>:</i>
<p>I give up, you win man. thanks for your help.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>I am trying to get a patch through in <a href="https://issues.jenkins.io/browse/JENKINS-16376" title="BuildStepMonitor.BUILD makes concurrent builds wait, could be changed to BuildStepMonitor.NONE?" class="issue-link" data-issue-key="JENKINS-16376"><del>JENKINS-16376</del></a>; please push for changes in other affected plugins.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=byronbrummer">byronbrummer</a>:</i>
<p>I have to wholeheartedly agree with surya et al: There needs to be a (non-plugin) switch for this.  Perhaps not a global switch (although, why not?), but at an absolute minimum a per-job switch.  This is a Jenkins-wide issue and giving plugins enough rope to hang the users does not strike me as the best process.</p>

<p>The only "workaround" really is to throw out the entire Post-Build functionality and recreated it from whole cloth as Build Steps.  That is really what this bug means: The entire Post-Build feature must be considered suspect and unreliable.  At the very minimum there should be a huge user warning around the Concurrent Build check box telling users that Post-Build actions may likely serialize their jobs anyway and should be avoided.</p>

<hr />

<p>It should also be noted that Jenkins jobs are leveraged for MUCH more than simply "pulling code from an SCM and compiling it" type jobs.  There are a huge number of utility tasks around a typical development lifecycle (deployments, environment setups and refreshes, data jobs, etc) that fit very well into Jenkins.  In the real world I see such utility jobs far, far outnumber "real" build jobs.  I mention this because of all the debate above about Git vs Subversion workflows and the like: It doesn't matter...in real world Jenkins usage it's common to find that MOST jobs aren't calling SCM <b>at all</b> because they aren't "buliding" anything, rendering all such workflow debates mute.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Build step monitors are an aspect of <em>any</em> build step, including both “builders” and “post-build steps” (recorders and notifiers), so that is an artificial distinction. Prior to the introduction of concurrent builds, plugins were free to assume that a build with a lower number than the current one was in fact completed—and some did in fact make that assumption. Monitors were introduced to ensure full binary compatibility. All plugins with build steps need to be updated to specify the monitor style they actually require, which is usually none. This has to be done on a case-by-case basis, while reviewing the plugin code for calls such as <tt>getPreviousBuild</tt> and ensuring that the logic continues to make sense when this build might be finishing well before earlier builds.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=notagain">notagain</a>:</i>
<p>Got bitten by the same thing. Now planning to move the post build stuff into an additional last step of the build job. <br/>
Only problem is that I need to get data from the slave to master before that. And that is also a Post build action (I use Copy-To-Slave plugin). Any way I can do it as a part of build job only?</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=reksiopl">reksiopl</a>:</i>
<p>To resolve this problem use xplugin as an intermediary to nunit/junit.  </p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Tim-Clifford<br/>
Path:<br/>
 src/main/java/jenkins/plugins/slack/ActiveNotifier.java<br/>
<a href="http://jenkins-ci.org/commit/slack-plugin/832e7b50028ca086dd49311ce454e7a19c9a3af9" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/slack-plugin/832e7b50028ca086dd49311ce454e7a19c9a3af9</a><br/>
Log:<br/>
  Change .getPreviousBuild() to allow for higher concurrency</p>

<p>Jenkins changed getPreviousBuild() to halt jobs if a previous<br/>
build hasn't finished, meaning jobs of variable runtime end up<br/>
being serialized:</p>

<p><a href="https://issues.jenkins-ci.org/browse/JENKINS-9913?focusedCommentId=184188&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-184188" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-9913?focusedCommentId=184188&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-184188</a></p>

<p>This change switches to the latest completed build, allowing<br/>
this plugin to work with jobs of variable length.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ericc">ericc</a>:</i>
<p>My apologies if this is the wrong please to ask this but I really did look for the right place and couldn't find it - is there a schedule for when this fix will be released? I am running into this issue and would benefit from the fix.</p>

**Comment by timja-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ericc" class="user-hover" rel="ericc">ericc</a> there are distinct issues for various plugins.</p>
